guide_drafts/model-derives.md

```diff
@@ -0,0 +1,769 @@
+    pub last_name: String,
+    pub email: String,
+}
+```
+
+*src/main.rs*
+```rust
+// externs, database connection, and using statement code omitted
+
+fn main() {
+    // The following will return all Users as a QueryResult<Vec<User>>
+    let query_result = users.load::<User>(&db_connection)
+      .expect("Error loading users");
+}
+```
+
+When deriving Queryable for a model, the struct must have all fields defined. Queryable's associated type, Row,  implements `FromSqlRow`, which provides the `fields_needed()` method. That method returns the number of fields to be consumed by the row. If we were to comment out all the three `String` fields on our User struct, we would see the following error.
+
+```rust
+error[E0277]: the trait bound `(i32,): diesel::types::FromSqlRow<(diesel::types::Integer, diesel::types::Text, diesel::types::Text, diesel::types::Text), _>` is not satisfied
+  --> src/bin/main.rs:21:28
+   |
+21 |     let query_result = users.load::<User>(&db_connection).expect("Error loading users");
+   |                            ^^^^ the trait `diesel::types::FromSqlRow<(diesel::types::Integer, diesel::types::Text, diesel::types::Text, diesel::types::Text), _>` is not implemented for `(i32,)`
+   |
+   = help: the following implementations were found:
+             <(A,) as diesel::types::FromSqlRow<(SA,), DB>>
+   = note: required because of the requirements on the impl of `diesel::Queryable<(diesel::types::Integer, diesel::types::Text, diesel::types::Text, diesel::types::Text), _>` for `diesel_demo_cli::models::User`
+   = note: required because of the requirements on the impl of `diesel::LoadQuery<_, diesel_demo_cli::models::User>` for `diesel_demo_cli::schema::users::table`
+```
+
+Notice the compiler is indicating a trait is not implmented for using the `.load()` method. FromSqlRow is expecting a tuple that looks like `(i32, std::string::String, std::string::String, std::string::String)`, but we currently only have a tuple consisting of `(i32,)`.
+
+This trait bound is illustrated below in the "hand-written" derive example.
+
+```rust
+
+// Implementing the Queryable trait by hand
+
+use diesel::query_source::Queryable;
+use diesel::backend::Backend;
+use diesel::types::{Integer, Text, HasSqlType, FromSqlRow};
+
+pub struct User {
+    pub id: i32,
+    pub first_name: String,
+    pub last_name: String,
+    pub email: String,
+}
+
+impl<ST, DB> Queryable<ST, DB> for User where
+    DB: Backend + HasSqlType<ST>,
+    (i32, String, String, String): FromSqlRow<ST, DB> {
+    
+    type Row = (i32, String, String, String);
+
+    fn build(row: &Self::Row) -> Self {
+        User {
+            id: row.0,
+            first_name: row.1,
+            last_name: row.2,
+            email: row.3,
+        }
+    }
+}
+
+
+```
+
+The advantages of using the derive annotation should be clear here, as the more columns you have, the more code you would have to write. You can thank Rust for this zero-cost-abstraction!
```

Comment:
"zero-cost abstraction" with only one dash.

Also: Is code generation/macros a zero-cost abstraction? https://blog.rust-lang.org/2015/05/11/traits.html quotes

> What you don’t use, you don’t pay for [Stroustrup, 1994]. And further: What you do use, you couldn’t hand code any better.

Seems ~~abstract~~ ambiguous to me in this case